docs/import/add_disaster.js

```diff
@@ -0,0 +1,220 @@
+  if (disaster === SENTINEL_NEW_DISASTER_VALUE) {
+    $('#new-disaster').show();
+    $('#selected-disaster').hide();
+  } else {
+    // TODO: display more disaster info including current layers etc. We
+    // might not want want to create these asset pickers every time we toggle
+    // between disasters because it requires ee calls. For now it's quite fast,
+    // will handle in follow up PRs if it begins to be problem.
+    $('#new-disaster').hide();
+    $('#selected-disaster').show();
+    return createStateAssetPickers(disasters.get(disaster));
+  }
+}
+
+/**
+ * Writes the given details to a new disaster entry in firestore. Fails if
+ * there is an existing disaster with the same details.
+ * @param {string} disasterId of the form <year>-<name>
+ * @param {Array<string>} states array of state (abbreviations)
+ * @return {Promise<boolean>} returns true after successful write to firestore.
+ */
+function writeDisaster(disasterId, states) {
+  if (disasters.has(disasterId)) {
+    setStatus('Error: disaster with that name and year already exists.');
+    return Promise.resolve(false);
+  }
+
+  disasters.set(disasterId, states);
+  toggleState(disasterId);
+
+  clearStatus();
+  const disasterOptions = $('#disaster > option');
+  // comment needed to quiet eslint on no-invalid-this rules
+  disasterOptions.each(/* @this HTMLElement */ function() {
+    if ($(this).val() > disasterId) {
+      $(createOptionFrom(disasterId)).insertBefore($(this));
+      return false;
+    }
+  });
+  $('#disaster').val(disasterId);
+  return getFirestoreRoot()
+      .collection('disaster-metadata')
+      .doc(disasterId)
+      .set({states: states})
+      .then(() => true);
+}
+
+/**
+ * Onclick function for the new disaster form. Writes new disaster to firestore,
+ * local disasters map and disaster picker. Doesn't allow name, year or states
+ * to be empty fields.
+ * @return {Promise<boolean>} resolves true if new disaster was successfully
+ *     written.
+ */
+function addDisaster() {
+  const year = $('#year').val();
+  const name = $('#name').val();
+  const states = $('#states').val();
+
+  if (!year || !name || !states) {
+    setStatus('Error: Disaster name, year, and states are required.');
+    return Promise.resolve(false);
+  }
+  if (isNaN(year)) {
+    setStatus('Error: Year must be a number.');
+    return Promise.resolve(false);
+  }
+  const disasterId = year + '-' + name;
+  return writeDisaster(disasterId, states);
+}
+
+// Needed for testing :/
+const emptyCallback = () => {};
+
+// TODO: add functionality for adding assets to disaster records from these
+// pickers.
+/**
+ * Requests all assets in ee directories corresponding to given states and
+ * displays them in pickers. Right now, selecting on those pickers doesn't
+ * actually do anything.
+ * @param {Array<string>} states e.g. ['WA']
+ * @return {Promise<void>} completes when we've finished filling all state
+ * pickers.
+ */
+function createStateAssetPickers(states) {
```

Comment:
Can we rephrase this function into two parts? One does the ee.listAssets and stores the data in the result of the Promise. Then another function takes the Promise and puts the data into the page. That way, the result of the Promise can be cached, removing the performance issue of listing assets when you're switching back and forth, and the two functions can be tested independently.

Reply:
Nice done.